pokedex/simulator/pokedex/src/simulator.rs

```diff
@@ -359,3 +335,314 @@ impl Drop for SimulatorHandle {
+    }
+
+    fn set_enable(&mut self) {
+        self.0.store(true, Ordering::Release);
+    }
+
+    fn set_value(&mut self, val: [u8; 4]) {
+        let v = i32::from_le_bytes(val);
+        self.1.store(v, Ordering::Release);
+    }
+
+    fn get_value(&self) -> [u8; 4] {
+        let v = self.1.load(Ordering::Acquire);
+        v.to_le_bytes()
+    }
+}
+
+impl BusInfo {
+    pub fn from_device_config<P: AsRef<Path> + Debug>(p: P) -> miette::Result<Self> {
+        let content = std::fs::read_to_string(&p)
+            .into_diagnostic()
+            .wrap_err_with(|| format!("fail read path: {:?}", &p))?;
+
+        let filepath = p.as_ref().to_string_lossy();
+        let configuration: Vec<AddressSpaceDescNode> = knuffel::parse(&filepath, &content)?;
+
+        let mut segments = Vec::new();
+        let mut ic = None;
+        for node in configuration {
+            match node {
+                AddressSpaceDescNode::SRAM(sram) => {
+                    let naive_memory = NaiveMemory::new(sram.length as usize);
+                    let boxed: Box<dyn Addressable> = Box::new(naive_memory);
+                    segments.push(((sram.base..sram.base + sram.length), boxed))
+                }
+                AddressSpaceDescNode::MMIO(mmio_config) => {
+                    let (mmio_decoder, controllers) =
+                        MMIOAddrDecoder::try_build_from(mmio_config.mmap)?;
+                    ic = Some(controllers);
+                    let boxed: Box<dyn Addressable> = Box::new(mmio_decoder);
+                    segments.push((
+                        (mmio_config.base..mmio_config.base + mmio_config.length),
+                        boxed,
+                    ))
+                }
+            }
+        }
+
+        let mut overlapped_segment = None;
+        let mut unchecked_index: Vec<Range<u32>> =
+            segments.iter().map(|(index, _)| index.clone()).collect();
+        unchecked_index.sort_by_key(|range| range.start);
+        for window in unchecked_index.windows(2) {
+            let addr1 = &window[0];
+            let addr2 = &window[1];
+            if addr2.start < addr1.end {
+                overlapped_segment = Some(&window[1]);
+            }
+        }
+
+        if let Some(range) = overlapped_segment {
+            miette::bail!(
+                "Address space with offset={:#010x} length={:#010x} overlapped previous address space",
+                range.start,
+                range.end - range.start
+            )
+        }
+
+        Ok(Self {
+            interrupt_controllers: ic.unwrap(),
+            bus_bridge: BusBridge {
+                address_space: segments,
+            },
+        })
+    }
+}
+
+#[test]
+fn test_address_space() -> miette::Result<()> {
+    let _ = BusInfo::from_device_config("./assets/devices.kdl")?;
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum BusError {
+    OutOfMemory {
+        id: &'static str,
+        addr: u32,
+        len: u32,
+    },
+    Illegal,
+}
```

Comment:
1. OutOfMemory means the system does not have enough memory to perform malloc. It's not the right term here.
2. When is 'Illegal' intended to use?